### ISSUE TITLE
Line editing and history broken in some terminals

### ISSUE THREAD

**Issue by davidkennedy-examtime** [nixme/jazz_hands#1]
I've switched to using jazz_hands on Ruby 1.9.3 on Ubuntu 12. Default install, no customisations.

When I start a rails console everything looks correct. Output is awesome.

However, when I start typing, about 2/3 of the way across the screen the text suddenly wraps and overwrites the line in place - typically how an app would look when it doesn't understand the terminal setting. This happens before the terminal width is reached. Once this has happened the user is typing blind as the placement of characters is unpredictable.

Another way to trigger it is to use the command history and try to edit a previous line.

I've fiddled with terminal settings, sizes etc but have seen no major differences. It is repeatable on all the Ubuntu machines here. It happens on bash, zsh _etc_. 

On a Debian box, it's less severe,. but the line wraps and overwrites itself. 

On an OS X box, it's similar to Debian. Line wrapping is broken and repeated use of history moves the prompt across the screen unpredictably.

**Comment by stephenprater** [comment]
I've got the same problem.  It's in the custom prompt.   I undid the custom prompt by putting `Pry.config.prompt = Pry::DEFAULT_PROMPT` in the `.pryrc` file.

**Comment by limratana** [comment]
I am encountering this problem running rails console in an OS X terminal as well.  Even if I make the terminal very wide, some place along the way, the line wraps onto itself, clearing out what was there, making it difficult to read what you are writing.  If you arrow up to bring up the last command, and it is a long command, then you see the line and it wraps over itself.  I am not sing jazz_hands.

**Comment by shiftb** [comment]
I have the same problem. Stephenprater's suggestion fixed it.

**Comment by nixme** [comment]
Finally figured this issue out. The terminal codes used to make colored prompts make most readline libraries incorrectly calculate line length (see https://github.com/pry/pry/issues/493). GNU readline (and rb-readline) however support wrapping the control characters with `\001` and `\002` to indicate they shouldn't be counted (see http://stackoverflow.com/questions/8806643/colorized-output-breaks-linewrapping-with-readline/8916332#8916332). So I added support for that. Unfortunately, OS X's libedit readline-wrapper (which most rubies will compile against by default) doesn't recognize them. So I now disable prompt coloring for libedit. See a8ee9ff59dd60747da7427471666d29423644654.

Long story short, jazz_hands 0.4 now tries to work well across most readline situations. If you still encounter issues, please file a new issue. Thanks!
